### PR TITLE
rewrite to not have the option to not have global config, added the c…

### LINKED ISSUE #8: Return Client

### PR DESCRIPTION
**Pull request by olokken**
…onfiguration as singleton in the demos

### CODE REVIEW

File: src/Vipps.net/Services/LoginService.cs [hunk 1]
```diff
@@ -1,21 +1,50 @@
...
+        }
+        
+        public string GetStartLoginUri(StartLoginURIRequest startLoginUriRequest, CancellationToken cancellationToken = default)
         {
-            string startLoginUri = $"{VippsConfiguration.BaseUrl}/access-management-1.0/access/oauth2/auth" +
-                                   $"?client_id={VippsConfiguration.ClientId}" +
+            var options = _vippsConfigurationProvider.GetConfiguration(); 
+            var baseUrl = options.UseTestMode 
+                ? "https://apitest.vipps.no" 
+                : "https://api.vipps.no";
```
Comment (aslaksm) [inline src/Vipps.net/Services/LoginService.cs]: Same logic in `VippsHttpClient`, can refactor it out somewhere

File: src/Vipps.net/Services/LoginService.cs [hunk 2]
```diff
@@ -1,21 +1,50 @@
...
-using System.Security.Cryptography;
 using System.Threading;
 using System.Threading.Tasks;
-using Polly.Retry;
 using Vipps.net.Infrastructure;
 using Vipps.net.Models.Login;
 
 namespace Vipps.net.Services
 {
+    public interface IVippsLoginService
```
Comment (aslaksm) [inline src/Vipps.net/Services/LoginService.cs:11]: Imo should move these interfaces to their own file and remove `Vipps` from name so they're in line with the class names.

File: src/Vipps.net/Infrastructure/IVippsConfigurationProvider.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Vipps.net.Infrastructure
+{
+    public interface IVippsConfigurationProvider
```
Comment (aslaksm) [inline src/Vipps.net/Infrastructure/IVippsConfigurationProvider.cs]: Any reason why we can't just use the `VippsConfigurationOptions` directly? What does the Provider abstraction provide?

File: src/Vipps.net.AspCoreDemo/Controllers/LoginController.cs [hunk 4]
```diff
@@ -50,22 +52,22 @@ public async Task<InitCibaResponse> InitCiba()
             InitCibaRequest initCibaRequest = new InitCibaRequest
             {
                 Scope = "openid email name phoneNumber", 
-                PhoneNumber = "47375750", 
+                PhoneNumber = "40285760",
```
Comment (aslaksm) [inline src/Vipps.net.AspCoreDemo/Controllers/LoginController.cs]: Is this a real phone number?

**Review by aslaksm — COMMENTED** [review]
This looks really nice  🚀  The user experience seems really nice and from what I can see the refactor only touches what it needs to. Nice work 🙌

File: src/Vipps.net/Services/LoginService.cs (hunk 2 above)
Comment (olokken) [reply src/Vipps.net/Services/LoginService.cs:11]: Agree that they should have the same name, added "Vipps" to the classnames.

File: src/Vipps.net/Infrastructure/IVippsConfigurationProvider.cs (hunk 3 above)
Comment (olokken) [reply src/Vipps.net/Infrastructure/IVippsConfigurationProvider.cs]: Removed the abstraction, are now using the VippsConfigurationOptions directly

File: src/Vipps.net/Services/LoginService.cs (hunk 1 above)
Comment (olokken) [reply src/Vipps.net/Services/LoginService.cs]: Refactored to a new class in the Helpers folder

File: src/Tests/Vipps.net.Tests/AccessTokenCacheServiceTests.cs [hunk 5]
```diff
@@ -11,30 +11,33 @@ public class AccessTokenCacheServiceTests
         [TestMethod]
         public void Can_Retrieve_Saved_Valid()
         {
+            AccessTokenCacheService accessTokenCacheService = new AccessTokenCacheService();
```
Comment (aslaksm) [inline src/Tests/Vipps.net.Tests/AccessTokenCacheServiceTests.cs]: `var` 😏

File: src/Vipps.net.AspCoreDemo/Controllers/LoginController.cs [hunk 6]
```diff
@@ -29,7 +31,7 @@ public string GetStartLoginURI()
                 AuthenticationMethod = AuthenticationMethod.Basic
```
Comment (haakongarfjell) [inline src/Vipps.net.AspCoreDemo/Controllers/LoginController.cs]: Burde denne også ta inn som parameter i stedet for body?

File: src/Vipps.net.AspCoreDemo/Controllers/LoginController.cs [hunk 7]
```diff
@@ -50,22 +52,22 @@ public async Task<InitCibaResponse> InitCiba()
             InitCibaRequest initCibaRequest = new InitCibaRequest
             {
                 Scope = "openid email name phoneNumber",
```
Comment (haakongarfjell) [inline src/Vipps.net.AspCoreDemo/Controllers/LoginController.cs:53]: openid må uansett være en del av scope så burde ikke trenge å legge inn selv

**Review by aslaksm — APPROVED** [review]
:shipit:

**Review by haakongarfjell — APPROVED** [review]

**Pull request merged**